Mathlib/Topology/Algebra/RestrictedProduct.lean

```diff
@@ -0,0 +1,769 @@
+
+end cofinite
+
+end Topology
+
+section Compatibility
+/-!
+## Compatibility properties between algebra and topology
+-/
+
+variable {S S' : ι → Type*} -- subobject types
+variable [Π i, SetLike (S i) (R i)] [Π i, SetLike (S' i) (R' i)]
+variable {A : Π i, S i} {A' : Π i, S' i}
+variable {T : Set ι} {𝓕 : Filter ι}
+variable [Π i, TopologicalSpace (R i)] [Π i, TopologicalSpace (R' i)]
+
+section general
+
+@[to_additive]
+instance [Π i, Inv (R i)] [∀ i, InvMemClass (S i) (R i)] [∀ i, ContinuousInv (R i)] :
+    ContinuousInv (Πʳ i, [R i, A i]_[𝓕]) where
+  continuous_inv := by
+    rw [continuous_dom]
+    intro T hT
+    haveI : ContinuousInv (Πʳ i, [R i, A i]_[𝓟 T]) :=
+      isEmbedding_coe_of_principal.continuousInv fun _ ↦ rfl
+    exact (continuous_inclusion hT).comp continuous_inv
+
+@[to_additive]
+instance {G : Type*} [Π i, SMul G (R i)] [∀ i, SMulMemClass (S i) G (R i)]
+    [∀ i, ContinuousConstSMul G (R i)] :
+    ContinuousConstSMul G (Πʳ i, [R i, A i]_[𝓕]) where
+  continuous_const_smul g := by
+    rw [continuous_dom]
+    intro T hT
+    haveI : ContinuousConstSMul G (Πʳ i, [R i, A i]_[𝓟 T]) :=
+      isEmbedding_coe_of_principal.continuousConstSMul id rfl
+    exact (continuous_inclusion hT).comp (continuous_const_smul g)
+
+end general
+
+section principal
+
+@[to_additive]
+instance [Π i, Mul (R i)] [∀ i, MulMemClass (S i) (R i)] [∀ i, ContinuousMul (R i)] :
+    ContinuousMul (Πʳ i, [R i, A i]_[𝓟 T]) :=
+  let φ : Πʳ i, [R i, A i]_[𝓟 T] →ₙ* Π i, R i :=
+  { toFun := (↑)
+    map_mul' := fun _ _ ↦ rfl }
+  isEmbedding_coe_of_principal.continuousMul φ
+
+@[to_additive]
+instance {G : Type*} [TopologicalSpace G] [Π i, SMul G (R i)] [∀ i, SMulMemClass (S i) G (R i)]
+    [∀ i, ContinuousSMul G (R i)] :
+    ContinuousSMul G (Πʳ i, [R i, A i]_[𝓟 T]) :=
+  isEmbedding_coe_of_principal.continuousSMul continuous_id rfl
+
+@[to_additive]
+instance [Π i, Group (R i)] [∀ i, SubgroupClass (S i) (R i)] [∀ i, IsTopologicalGroup (R i)] :
+    IsTopologicalGroup (Πʳ i, [R i, A i]_[𝓟 T]) where
+
+instance [Π i, Ring (R i)] [∀ i, SubringClass (S i) (R i)] [∀ i, IsTopologicalRing (R i)] :
+    IsTopologicalRing (Πʳ i, [R i, A i]_[𝓟 T]) where
+
+end principal
+
+section cofinite
+
+theorem nhds_zero_eq_map_ofPre [Π i, Zero (R i)] [∀ i, ZeroMemClass (S i) (R i)]
+    (hAopen : ∀ i, IsOpen (A i : Set (R i))) (hT : cofinite ≤ 𝓟 T) :
+    (𝓝 (inclusion R (fun i ↦ A i) hT 0)) = map (inclusion R (fun i ↦ A i) hT) (𝓝 0) :=
+  nhds_eq_map_inclusion hAopen hT 0
+
+theorem nhds_zero_eq_map_structureMap [Π i, Zero (R i)] [∀ i, ZeroMemClass (S i) (R i)]
+    (hAopen : ∀ i, IsOpen (A i : Set (R i))) :
+    (𝓝 (structureMap R (fun i ↦ A i) cofinite 0)) =
+      map (structureMap R (fun i ↦ A i) cofinite) (𝓝 0) :=
+  nhds_eq_map_structureMap hAopen 0
+
+-- TODO: Make `IsOpen` a class like `IsClosed` ?
+variable [hAopen : Fact (∀ i, IsOpen (A i : Set (R i)))]
+variable [hAopen' : Fact (∀ i, IsOpen (A' i : Set (R' i)))]
```

Comment:
This line is never used. In fact is it the case that there's just one result in the whole 750+-line file which uses R' and A' even though they're initialised globally?

Reply:
This may be the case. I don't know what the policy is here, I know some people claim that you should always have your data at the section level and only shuffle hypotheses on the lemma level. But I agree this case is a bit extreme, so I'll remove the global `R'` and `A'`.